barretenberg/cpp/src/barretenberg/protogalaxy/protogalaxy_prover.hpp

```diff
@@ -336,39 +381,104 @@ template <class ProverInstances_> class ProtoGalaxyProver_ {
+                if constexpr (disable_optimisation) {
+                    accumulate_relation_univariates(
+                        thread_univariate_accumulators[thread_idx],
+                        extended_univariates[thread_idx],
+                        instances.relation_parameters, // these parameters have already been folded
+                        pow_challenge);
+                } else {
+                    accumulate_relation_univariates(
+                        thread_univariate_accumulators[thread_idx],
+                        extended_univariates[thread_idx],
+                        instances.optimised_relation_parameters, // these parameters have already been folded
+                        pow_challenge);
+                }
             }
         });
+        if constexpr (disable_optimisation) {
+            Utils::zero_univariates(univariate_accumulators);
+            // Accumulate the per-thread univariate accumulators into a single set of accumulators
+            for (auto& accumulators : thread_univariate_accumulators) {
+                Utils::add_nested_tuples(univariate_accumulators, accumulators);
+            }
 
-        // Accumulate the per-thread univariate accumulators into a single set of accumulators
-        for (auto& accumulators : thread_univariate_accumulators) {
-            Utils::add_nested_tuples(univariate_accumulators, accumulators);
+            return batch_over_relations(univariate_accumulators, instances.alphas);
+        } else {
+            Utils::zero_univariates(optimised_univariate_accumulators);
+            // Accumulate the per-thread univariate accumulators into a single set of accumulators
+            for (auto& accumulators : thread_univariate_accumulators) {
+                Utils::add_nested_tuples(optimised_univariate_accumulators, accumulators);
+            }
+
+            // Convert from optimised version to non-optimised
+            deoptimise_univariates(optimised_univariate_accumulators, univariate_accumulators);
+            //  Batch the univariate contributions from each sub-relation to obtain the round univariate
+            return batch_over_relations(univariate_accumulators, instances.alphas);
         }
-        // Batch the univariate contributions from each sub-relation to obtain the round univariate
-        return batch_over_relations(univariate_accumulators, instances.alphas);
+    }
+
+    /**
+     * @brief Convert univariates from optimised form to regular
+     *
+     * @details We need to convert before we batch relations, since optimised versions don't have enough information to
+     * extend the univariates to maximum length
+     *
+     * @param optimised_univariate_accumulators
+     * @param new_univariate_accumulators
+     */
+    static void deoptimise_univariates(const OptimisedTupleOfTuplesOfUnivariates& optimised_univariate_accumulators,
+                                       TupleOfTuplesOfUnivariates& new_univariate_accumulators
```

Comment:
can we not compute the univariates, use only the information we need in combiner computation and then use the entire univariate in batching rather than optimising and deoptimising?

Reply:
That's what we are doing, but we need to switch the arithmetic back to normal univariate before we do extensions and batching